### ISSUE TITLE
Rewrite format-reference to use citeproc-el

### ISSUE THREAD

**Issue by bdarcus** [bdarcus/citar#245]
If replacing bibtex-completion function for this, might as well use citeproc, since it would provide much more flexibility (mode-specific output, thousands of CSL styles for formatting specification, etc. etc.)..

Also, citeproc is needed anyway if and when we implement things like style previews in `oc-bibtex-actions`.

Downside is an additional dependency.

**Comment by localauthor** [comment]
Simple question: is the intent here to replace `bibtex-completion-insert-reference`?

If so, would the attendant flexibility you mention include allowing references to be formatted in different styles (not just APA, but, Chicago, MLA, Harvard, etc.)? Inquiring humanists want to know ;)

**Comment by bdarcus** [comment]
Yes, and if integrating citeproc-el, yes.

But while I have you, what's your use case for this? Are you using org here primarily, for example?

**Comment by localauthor** [comment]
Yes, I write manuscripts and take readings notes in org.

In manuscripts I use org-cite, so I don't need `insert-reference`, because it's enough to have everything formatted upon export. But I don't export my reading notes, so if I want to include a formatted list of sources in a note---in an annotated bibliography, say---`insert-reference` is a quick way to do that. I'm just not used to APA, so being able to use Chicago would be nice.

**Comment by bdarcus** [comment]
I'm wondering if for annotated bibs and CVs if there's a better way using `oc-csl|?

I think @titaniumbones had asked about this, and found a CSL style designed for this, where you could just use citations, and they would be formatted as references.

**Comment by titaniumbones** [comment]
oops, sorry for the delay. I did find a CSL style for this -- it's apa-cv.csl, and I can't quite remember where I found it but it wasn't hard. I will make a Chicago version of it too, when I have a sec. If it seems useful to other people, I can put both up in github somewhere once I dig out a bit.

**Comment by bdarcus** [comment]
Maybe we could move this command out of the main file, and create a new `bibtex-actions-citeproc.el` file where a revised version could go.

`oc-bibtex-actions` could then require that.

WDYT @aikrahguzar?

**Comment by bdarcus** [comment]
> oops, sorry for the delay. I did find a CSL style for this -- it's apa-cv.csl, and I can't quite remember where I found it but it wasn't hard. I will make a Chicago version of it too, when I have a sec. If it seems useful to other people, I can put both up in github somewhere once I dig out a bit.

It was occurred to me it makes at this point to encourage more distributed style development and hosting. That was the original intention.

Let me know if you post it on a repo.

**Comment by bdarcus** [comment]
There already is a chicago style for annotated bibs, though I haven't tested or looked at it closely.

https://github.com/citation-style-language/styles/blob/master/chicago-annotated-bibliography.csl

**Comment by andersjohansson** [comment]
Also needed this. My use-case: quickly inserting references in emails.
Hacked together this function for now:
```emacs-lisp
(setq citar-insert-reference-function #'aj/citar-get-csl-references)

(defun aj/citar-get-csl-references (keys)
  (let ((proc (citeproc-create "/home/aj/.zotero-profile/styles/apa.csl"
                               ;; (lambda (_)
                               ;;   (cl-loop for (k . v) in keys
                               ;;            collect
                               ;;            (list k v)))
                               (citeproc-hash-itemgetter-from-any citar-bibliography)
                               (org-cite-csl--locale-getter))))
    (citeproc-add-uncited (mapcar #'car keys) proc)
    (insert (car (citeproc-render-bib proc 'plain)))))
```
I expected citeproc to be able to use the data from citar for the itemgetter (the lambda above) but that didn’t seem to work, so for now it’s an extra lookup in the bib-file (not very efficient).

By the way, the docstring for `citar-insert-reference-function` is "A function that takes a list of (KEY . ENTRY), and returns formatted references.", which is a bit confusing since it shouldn’t return the references but insert, as stated by the name (was confused by this for a while). But I see that this may change anyway: #448.

**Comment by bdarcus** [comment]
> I expected citeproc to be able to use the data from citar for the itemgetter (the lambda above) but that didn’t seem to work ...

@andras-simonyi - any input here?

**Comment by bdarcus** [comment]
On #448, I like the idea of adding a `citar-copy-reference` and making that available at point, per https://github.com/bdarcus/citar/issues/448#issuecomment-981818707.

So it seems that plus maybe `citar-citeproc` would be a good resolution for this issue.

**Comment by andersjohansson** [comment]
> I expected citeproc to be able to use the data from citar for the itemgetter (the lambda above) but that didn’t seem to work ...

But looking into it a bit more, the data in `citar--candidates-cache` is probably not enough for citeproc, unless we add more fields with `citar-additional-fields`. But my use with mainly org means I also have a populated `org-cite-basic--bibliography-cache`. It would of course be good to use available caches rather than reading from file again with `citeproc-hash-itemgetter-from-any`, but I suppose it is a bit difficult to come up with a good solution for different use cases.

**Comment by andersjohansson** [comment]
And to explain what happens: What I do get out with the data from `citar--candidates-cache` from the (commented) function above is just "(n.d.).":

Trying to insert one reference and doing some edebugging on `citeproc-render-bib` gives itemdata like this:
```
(citeproc-itemdata :varvals ((citation-number . "1") (label . "page") ((id . "traxler.etal-2018-gender.fairness.force") (author . "Traxler, Adrienne and Henderson, Rachel and Stewar...") (file . "/home/aj/bibliotek/Traxler et al - 2018 - Gender f...") (issued . "2018") (title . "Gender fairness within the Force Concept Inventory") (type . "article-journal"))) :rawcite nil :rawbibitem nil :rc-uptodate nil :sort-key (#1="" "0" #1# #1# #1#) :occurred-before nil :disamb-pos nil :subbib-nos nil)
```

Which for some reason is boils down to this `(((bib-item-no . "1")) "(" (((rendered-var . issued)) "n.d.") ").")`, rendered as "(n.d.)."

**Comment by bdarcus** [comment]
> What I do get out with the data from `citar--candidates-cache` from the (commented) function above is just "(n.d.).":

I haven't read this carefully (in a hurry), but is this possibly a bug in citeproc? See this thread on reddit today.

https://www.reddit.com/r/orgmode/comments/r7p4hs/orgexport_exports_without_date_information_with/

**Comment by andersjohansson** [comment]
> I haven't read this carefully (in a hurry), but is this possibly a bug in citeproc? See this thread on reddit today.
> 
> https://www.reddit.com/r/orgmode/comments/r7p4hs/orgexport_exports_without_date_information_with/
> 
Yes possibly. I keep my data in a csl-json file exported from zotero with BBT (the parsing is so much quicker). I’ve had similar problems with mismatching field names causing some missing output for org exports with citeproc. For this case I get only “n.d.” though. Could it be that no field name matches here?

**Comment by bdarcus** [comment]
CSL uses `issued`, but the default templates include that, so it should be included.

Did you modify those templates?

**Comment by andersjohansson** [comment]
No, I haven‘t changed the templates. This is the mentioned entry in `citar--candidates-cache`
```
(#("Traxler, Henderson, Stewart, Stewart, Papak, Lindell     2018     Gender fairness within the Force Concept Inventory           traxler.etal-2018-gender.fairness.force                                                                                      has:files    traxler.etal-2018-gender.fairness.force" 0 30
	(face citar-highlight)
	30 52
	(invisible t face citar-highlight)
	52 114
	(face citar-highlight)
	114 116
	(invisible t face citar-highlight)
	117 142
	(face citar)
	142 166
	(invisible t face citar)
	166 250
	(face citar)
	251 304
	(invisible t))
  #521="traxler.etal-2018-gender.fairness.force"
  (id . #521#)
  (author . "Traxler, Adrienne and Henderson, Rachel and Stewart, John and Stewart, Gay and Papak, Alexis and Lindell, Rebecca")
  (file . "/home/aj/bibliotek/Traxler et al - 2018 - Gender fairness within the Force Concept Inventory.pdf")
  (issued . "2018")
  (title . "Gender fairness within the Force Concept Inventory")
  (type . "article-journal"))
```

**Comment by bdarcus** [comment]
And `issued` turns into `n.d.` on export?

The data's there, so is this likely a citeproc-el issue? Or something with how you're using it?

**Comment by andras-simonyi** [comment]
Apologies for reacting only now, unfortunately, I have been extremely busy lately. If the task is simply to render an isolated item then it's better to use `citeproc-render-item`. Or do you have several items and they required to be sorted? As for using the citar cache this should be possible but depends on the format -- `citeproc-el` expects the item data to be in parsed csl-json format. Looking at the example above this means that the date and name fields would have to be converted. If they are stored in BibTeX format then `citeproc-el` contains some convenient functions for doing this.

**Comment by bdarcus** [comment]
Forgot to link this; closed via #455.
